### DISCUSSION TITLE
RP2350 board: What is used in development / what would you recommend ?

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#156] [General]
Hi,
Just out of curiosity I want an RP2350 board with RAM (PSRAM and including other supported kind of RAMs) and WiFi/BLE for
experimenting with the new features of Zeptoforth :)
But I want to prevent buying a board which is not yet supported or "too much a special case"... ;)
I would order it from germany...

Any hints or recommandations?

Cheers!
:)

**Comment by tabemann** [comment]
On Sat, Nov 9, 2024 at 6:31 AM mcctuxic ***@***.***> wrote:

> Hi,
> Just out of curiosity I want an RP2350 board with RAM (PSRAM and including
> other supported kind of RAMs) and WiFi/BLE for
> experimenting with the new features of Zeptoforth :)
> But I want to prevent buying a board which is not yet supported or "too
> much a special case"... ;)
> I would order it from germany...
>
> Any hints or recommandations?
>
> Cheers!
> :)
>
> If you want to try out PSRAM, I would recommend the Pimoroni Pico Plus 2.
Note that currently zeptoforth does not support all its flash out of the
box, as zeptoforth for the RP2350 only currently uses 4 MB of flash (2 MB
for the flash dictionary, 2 MB for blocks and filesystems); zeptoforth
would have to be modified to use the full 8 MB of flash it provides.

As for WiFi/BLE, I would highly recommend the Raspberry Pi Pico 2 W when it
comes out, but it hasn't come out yet. zeptoforth should support it as soon
as it is out, as it is supposed to use the same wireless module as the Pico
W. Also note that zeptoforth only supports WiFi and not BLE. However, I
don't expect it to have any PSRAM onboard.

There is the Challenger+ RP2350 WiFi6/BLE5
https://ilabs.se/product/challenger-rp2350-wifi-ble/, which has 8 MB PSRAM
and a ESP32-C6 wireless module, but mind you you are on your own with it to
implement networking, because zeptoIP doesn't and won't support it.
(zeptoIP only supports interfaces that directly expose Ethernet frames to
the user, which the ESP32-C6 module does not do.) I can't find any other
RP2350 boards that currently have WiFi support, much the less WiFi support
plus PSRAM.

Travis

**Reply by mcctuxic** [reply]
Hi Travis,

Thanks a lot for the information - very appreciated! :)

I saw the Pimoroni Pico Plus 2 **W**
https://shop.pimoroni.com/products/pimoroni-pico-plus-2-w?variant=42182811942995
(but it is currently out of stock) - would that one become supported?
What do you think about it?

> as zeptoforth for the RP2350 only currently uses 4 MB of flash (2 MB
for the flash dictionary, 2 MB for blocks and filesystems); zeptoforth
would have to be modified to use the full 8 MB of flash it provides.

(WITHOUT PUTTING ANY PRESSURE ON YOU, TRAVIS...it is just a question...)
Is this modification for using the complete flash something you have in mind
for near or not-so-near future of zeptoforth?

Cheers!
mcctuxic

**Comment by znmeb** [comment]
> There is the Challenger+ RP2350 WiFi6/BLE5
https://ilabs.se/product/challenger-rp2350-wifi-ble/, which has 8 MB PSRAM
and a ESP32-C6 wireless module, but mind you you are on your own with it to
implement networking, because zeptoIP doesn't and won't support it.

I have a couple of these. I haven't had a chance to experiment with the WiFI yet but there is a pretty good description of how to get to the WiFi module from Arduino C/C++ on this page: https://ilabs.se/challenger-rp2350-wifi-ble/. The hardware-level communication protocol is documented there and it should be accessible via user Forth code. There doesn't appear to be any need to add any words to the existing Zeptoforth builds.

I also have a couple of the Pimoroni Pico Plus 2 W boards, but the only working software for them currently is Pimoroni's MicroPython build. The Pico SDK doesn't support them directly yet. Neither does the Arduino Pico core.

**Comment by tabemann** [comment]
On Sat, Nov 9, 2024 at 10:48 AM mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> Thanks a lot for the information - very appreciated! :)
>
> I saw the Pimoroni Pico Plus 2 *W*
>
> https://shop.pimoroni.com/products/pimoroni-pico-plus-2-w?variant=42182811942995
> (but it is currently out of stock) - would that one become supported?
> What do you think about it?
>
In the very least, even if it were feasible, it would require writing a new
WiFi driver because it uses a different module for WiFi from the Raspberry
Pi Pico W. However, if it is one of those modules that internally implement
TCP/IP rather than exposing an Ethernet frame interface, then a whole new
API layer would have to be written for it. Of course, developing this
without an actual board on hand might be tricky.

> as zeptoforth for the RP2350 only currently uses 4 MB of flash (2 MB
> for the flash dictionary, 2 MB for blocks and filesystems); zeptoforth
> would have to be modified to use the full 8 MB of flash it provides.
>
> (WITHOUT PUTTING ANY PRESSURE ON YOU, TRAVIS...it is just a question...)
> Is this modification for using the complete flash something you have in
> mind
> for near or not-so-near future of zeptoforth?
>
> Cheers!
> mcctuxic
>
The problem with "officially" modifying zeptoforth to support the full 8 MB
of flash is that it would require making a new "platform" of builds, which
would even more time to my build process. However, it should be trivial to
make the source code changes necessary to make zeptoforth support an 8 MB
of flash so as to support 6 MB of block/filesystem space (it is less
trivial to make it support more than 2 MB of dictionary space, because then
you would have to expand the size of the "minidictionary" to make it
practical). Really all you have to do is make the following change in
src/rp230/forth/qspi.fs:

        $400000 constant QUADSPI_Size

to

        $800000 constant QUADSPI_Size

Then just rebuild zeptoforth yourself per the README.

Travis

**Comment by tabemann** [comment]
On Sat, Nov 9, 2024 at 1:21 PM M. Edward (Ed) Borasky <
***@***.***> wrote:

> There is the Challenger+ RP2350 WiFi6/BLE5
> https://ilabs.se/product/challenger-rp2350-wifi-ble/, which has 8 MB PSRAM
> and a ESP32-C6 wireless module, but mind you you are on your own with it to
> implement networking, because zeptoIP doesn't and won't support it.
>
> I have a couple of these. I haven't had a chance to experiment with the
> WiFI yet but there is a pretty good description of how to get to the WiFi
> module from Arduino C/C++ on this page:
> https://ilabs.se/challenger-rp2350-wifi-ble/. The hardware-level
> communication protocol is documented there and it should be accessible via
> Forth code. There doesn't appear to be any need to add any words to the
> existing Zeptoforth builds.
>
> I also have a couple of the Pimoroni Pico Plus 2 W boards, but the only
> working software for them currently is Pimoroni's MicroPython build. The
> Pico SDK doesn't support them directly yet. Neither does the Arduino Pico
> core.
>
> I'm personally waiting for the Raspberry Pi Pico 2 W because it should be
trivial to interface with zeptoIP as currently implemented.

Travis

**Reply by znmeb** [reply]
Would that also work for the Arduino Nano RP2040 Connect? https://docs.arduino.cc/hardware/nano-rp2040-connect/. I have a couple of those too.

**Comment by tabemann** [comment]
On Sat, Nov 9, 2024 at 5:22 PM M. Edward (Ed) Borasky <
***@***.***> wrote:

> Would that also work for the Arduino Nano RP2040 Connect?
> https://docs.arduino.cc/hardware/nano-rp2040-connect/. I have a couple of
> those too.
>
From looking at the Nina W102 module used by the Arduino Nano RP2040
Connect, it seems like the module implements TCP/IP internally, so you
couldn't use zeptoIP and would have to write your own layer to interface
with it.

Travis
